sdk/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs

```diff
@@ -61,6 +61,56 @@ runGammaUnderUpgrades Upgrading{ _past = pastAction, _present = presentAction }
     presentResult <- withReaderT _present presentAction
     pure Upgrading { _past = pastResult, _present = presentResult }
 
+checkUpgradeDependencies
+    :: Version
+    -> Bool
+    -> LF.Package
+    -> [LF.DalfPackage]
+    -> Maybe ((LF.PackageId, LF.Package), [(LF.PackageId, LF.Package)])
+    -> [Diagnostic]
+checkUpgradeDependencies _ _ _ _ Nothing = []
+checkUpgradeDependencies version shouldTypecheckUpgrades presentPkg presentDeps (Just ((_pastPkgId, pastPkg), pastDeps)) =
+    let package = Upgrading { _past = pastPkg, _present = presentPkg }
+        upgradingWorld = fmap (\package -> emptyGamma (initWorldSelf [] package) version) package
+    in
+    extractDiagnostics $
+        runGammaF upgradingWorld $
+            when shouldTypecheckUpgrades (checkUpgradeDependenciesM presentPkg presentDeps pastPkg pastDeps)
+
+checkUpgradeDependenciesM
+    :: LF.Package
+    -> [LF.DalfPackage]
+    -> LF.Package
+    -> [(LF.PackageId, LF.Package)]
+    -> TcUpgradeM ()
+checkUpgradeDependenciesM _presentPkg presentDeps _pastPkg pastDeps = do
+    let packageToNameVersion :: LF.Package -> (LF.PackageName, LF.PackageVersion)
+        packageToNameVersion LF.Package{packageMetadata = LF.PackageMetadata{packageName, packageVersion}} =
+            (packageName, packageVersion)
+        presentDepsMap =
+            HMS.fromList $ map (packageToNameVersion . extPackagePkg . dalfPackagePkg) presentDeps
+        pastDepsMap =
+            HMS.fromList $ map (packageToNameVersion . snd) pastDeps
+    let (_del, existingDeps, _new) = extractDelExistNew Upgrading { _past = pastDepsMap, _present = presentDepsMap }
+    forM_ (HMS.toList existingDeps) $ \(depName, depVersions) -> do
+        case LF.comparePackageVersion (_present depVersions) (_past depVersions) of
+          Left (FirstVersionUnparseable presentVersion) ->
```

Comment:
It feels like `comparePackageVersion` is maybe mixing concerns here? I would have expected `PackageVersion` to be a parsed version already, and `comparePackageVersion` to do only the comparing. Is it the only place where we have to actually consider versions as something other than a plain string?

I appreciate that changing the way we represent versions is out of scope of this PR. But maybe you could separate the parsing function from the comparison one?

Reply:
Good point - I also wondered what the newtype is good for if there's no validation done anyways that it's valid at decode.

I've moved the parsing within Either monad into `splitPackageVersion`, and refactored `comparePackageVersion` to use do notation so we lose case statements and other noise.